File: fs/projects/src/il/co/ilrd/linked_list/LinkedList.java
package il.co.ilrd.linked_list;

/*
 * Author: Shieber
 * Reviewer: Hila
 */

public class LinkedList {
    private Node head;
    
    public LinkedList() {
		head = new Node(null, null);
	}

    /**
     * Pushes element at the front of list.
     *
     * @param data Data to push into list.
     */
    public void pushFront(Object data) {
    	
    	Node oldHead = head;
    	head = new Node(oldHead, data);
    }

    /**
     * Removes and returns the first element of this list.
     *
     * @return The element at the front of this list, if list is empty null is returned.
     */
    public Object popFront() {
        
    	if (!isEmpty())
    	{
    		Object dataToRemove = head.data;
    		head = head.next;
    		return dataToRemove;
    	}
    	
    	return null;
    }

    /**
     * Checks if list is empty.
     *
     * @return true if empty, otherwise false.
     */
    public boolean isEmpty() {
        return (null == head.next);
    }

    /**
     * Get number of elements in the list.
     *
     * @return Number of elements in the list.
     */
    public int size() {
    	
    	int numOfNodes = 0;
    	Iterator iter = begin();
        
        while (iter.hasNext())
        {
        	iter.next();
        	++numOfNodes;
        }
        
    	return numOfNodes;
    }

    /**
     * Finds data in Linked List.
     *
     * @param data Data to find.
     * @return Iterator with found data, otherwise null.
     */
    public Iterator find(Object data) {
        Object curData = null;
        Iterator iter = begin();
        Iterator prevIter = begin();

        
        while (iter.hasNext())
        {
        	curData = iter.next();
        	if (data.equals(curData)) {
        		return prevIter;
        	}
        	else
        	{
        		prevIter.next();
        	}
        }
    	return null;
    }

    /**
     * @return Iterator to the first element
     */
    public Iterator begin() {
        return (new ListIterator(head));
    }

    private class ListIterator implements Iterator {
        private Node currentNode;
        
        public ListIterator(Node currentNode) {
			this.currentNode = currentNode;
		}

        @Override
        public boolean hasNext() {
            return (null != currentNode.next);
        }

        @Override
        public Object next() {
            Object data = currentNode.data;
            currentNode = currentNode.next;
        	return data;
        }
    }

    private class Node {
        private Object data;
        private Node next;

        private Node(Node next, Object data) {
            this.data = data;
            this.next = next;
        }
    }
}

File: anyvision/projects/linked_list/linked_list.js
/**
 * Represents a Singly linked list.
 */
class LinkedList {
  /**
   * Creates a Linked List.
   * @constructor
   */
  constructor() {
    this._size = 0;
    this._head = null;
  }

  /**
   * Checks if list is empty.
   * @returns {boolean} true if list is empty, false otherwise.
   */
  isEmpty() {
    return this._size === 0;
  }

  /**
   * Size of the list.
   * @returns {number} number of elements in the list.
   */
  size() {
    return this._size;
  }

  /**
   * Removes the last element added to the list.
   * @returns {undefined|data} - undefined if list is empty. | data of the removed element.
   */
  popFront() {
    if (this._head === null) {
      return undefined;
    }

    let removedData = this._head.data;
    this._head = this._head._next;
    --this._size;
    return removedData
  }

  /**
   * Insert data into the list.
   * @param {*} data - data to be inserted into the list.
   */
  pushFront(data) {
    this._head = new Node(data, this._head);
    ++this._size;
  }

  /**
   * Find data.
   * @param {*} data - data to find.
   * @param {function} cmpFunc - compare function for the comparator.
   */
  find(data, cmpFunc) {
    for (let node of this) {
      if (cmpFunc(node, data)) {
        return node;
      }
    }

    return undefined;
  }

  [Symbol.iterator]() {
    let cur = this._head;
    return {
      next: function () {
        if (cur !== null) {
          const data = cur._data;
          cur = cur._next;
          return { value: data, done: false };
        }

        return { value: undefined, done: true };
      },
    };
  }
}

class Node {
  constructor(data, next) {
    this._data = data;
    this._next = null;
  }
}

module.exports = { LinkedList };

